README.md
Java GUI实用小工具集 Common Gui Tools
=====================================

### 简介 ###
<a href="https://github.com/baishui2004/common_gui_tools" target="_blank">Common Gui Tools</a> 是用java编写，GUI界面的实用小工具集，1.5版分六个类别22个小工具。

### 安装 ###
1. Require Java 6.0+
2. 下载压缩包<a href="https://github.com/baishui2004/common_gui_tools/blob/master/dest/common_gui_tools-1.5.zip?raw=true">common_gui_tools-1.5.zip</a>，解压，Windows下双击start.bat运行，Mac或Linux下双击start.sh运行
3. 软件打开后通过菜单栏Character Tools、File Tools等选择要打开的工具

### 工具集索引 ###
#### 字符集工具集 ####
1. <a href="#Character Converter">编码转换：Character Converter</a>
2. <a href="#Encrypt And Decrypt">加密解密：Encrypt And Decrypt</a>
3. <a href="#Escape Character Tool">字符转义工具：Escape Character Tool</a>
4. <a href="#JUniversal Chardet">文本编码识别：JUniversal Chardet</a>
#### 文件工具集 ####
1. <a href="#File Digital Signature">计算文件数字签名：File Digital Signature</a>
2. <a href="#Folder And File Operate">文件(夹)查找操作：Folder And File Operate</a>
3. <a href="#JNotify Visual">文件(夹)变化监控：JNotify Visual</a>
4. <a href="#JODConverter Visual">OpenOffice文档转换：JODConverter Visual</a>
5. <a href="#Text File Split">文本文件切分：Text File Split</a>
#### 常用工具集 ####
1. <a href="#Code Formatter">Code格式化：Code Formatter</a>
2. <a href="#Password Generator">随机密码生成：Password Generator</a>
3. <a href="#QrCode Converter">二维码转换：QrCode Converter</a>
4. <a href="#Regex Tester">正则表达式验证：Regex Tester</a>
5. <a href="#Run Script">执行Script脚本：Run Script</a>
6. <a href="#Time Tool">时间工具：Time Tool</a>
7. <a href="#ZHConverter Visual">中文简体繁体互转：ZHConverter Visual</a>
#### 金融工具集 ####
1. <a href="#Interest Calculator">利息利率计算器：Interest Calculator</a>
#### 其他工具集 ####
1. <a href="#Color Tool">颜色工具：Color Tool</a>
2. <a href="#System Information">运行环境信息：System Information</a>
#### 过时工具集 ####
1. <a href="#Ant Script Auto Build">Ant脚本自动build：Ant Script Auto Build</a>
2. <a href="#Class Finder">Java类查找：Class Finder</a>
3. <a href="#Simple Webview">简易Webview：Simple Webview</a>

### 配置说明 ### 
```
1，本软件采用插件方式，22个小工具即是22个插件，插件配置文件夹conf，
   配置文件有：common_gui_tools.properties、tools.properties以及多个插件的配置；
2，相关配置说明参看各配置文件，可通过修改文件common_gui_tools.properties中属性CommonUseTools修改常用插件;
   默认加载插件Encrypt And Decrypt、Folder And File Operate、Time Tool、Code Formatter； 
3，可通过GUISkin属性修改软件皮肤，fontStyles系列属性修改显示字体。
```

### 开发说明 ### 
```
1，启动类：bs.tool.commongui.GuiMain，参见：src/main/MANIFEST.MF；
2，本地开发运行前先执行mvn clean package，以将src/main/sources下资源拷贝到target，否则运行时会找不到配置文件及图片；
```

### 捐助本项目 ### 
支持长远发展，感谢您的认可！
<br /><b>微信</b>  
![Donate weixin](./documention/images/donate/weixin.png)
<br /><b>支付宝</b>  
![Donate alipay](./documention/images/donate/alipay.jpg)

### 详细介绍 ### 
以下详细简绍每个小工具：
#### 字符集工具集 ####
<b>1. <a id="Character Converter">编码转换：Character Converter</a></b>  
(1)，编码：Encode String表单输入字符，点击右侧对应的Encode按钮，对输入字符进行编码，注意此时的编码类别是“二进制”、“八进制”、“十进制”、“十六进制”，对应的可以使用下面的Decode进行解码  
![Character Converter](./documention/images/Character%20Converter-1.png)  
(2)，乱码解码：选择编码类别“乱码解码”，比如在Big5表单中输入“趼睫”，点击右侧对应Decode按钮，GBK表单中解码出字符“字符”  
![Character Converter](./documention/images/Character%20Converter-2.png)

<b>2. <a id="Encrypt And Decrypt">加密解密：Encrypt And Decrypt</a></b>  
加密解密：默认字符集UTF-8，另可选其他常用字符集，前五种算法可解密，后五种算法不可逆  
![Encrypt And Decrypt](./documention/images/Encrypt%20And%20Decrypt.png)

<b>3. <a id="Escape Character Tool">字符转义工具：Escape Character Tool</a></b>  
HTML、XML、JAVA、JavaScript、CSV字符转义字符及还原，使用Apache Commons-lang的StringEscapeUtils类  
![Escape Character Tool](./documention/images/Escape%20Character%20Tool.png)

<b>4. <a id="JUniversal Chardet">文本编码识别：JUniversal Chardet</a></b>  
检测文件编码，识别准确率高（有一定的误差）  
![JUniversal Chardet](./documention/images/JUniversal%20Chardet.png)  
项目：<a href="https://code.google.com/p/juniversalchardet/" target="_blank">juniversalchardet</a>  
文档：<a href="http://www-archive.mozilla.org/projects/intl/UniversalCharsetDetection.html" target="_blank">Mozilla UniversalCharsetDetection</a>  
其他编码识别项目：<a href="http://sourceforge.net/projects/jchardet/" target="_blank">jchardet</a>&emsp;&nbsp;<a href="http://sourceforge.net/projects/cpdetector/" target="_blank">cpdetector</a>&emsp;&nbsp;<a href="http://wing.comp.nus.edu.sg/~tanyeefa/downloads/charsetdetectstreamreader/" target="_blank">Charset Detect Stream Reader</a>

#### 文件工具集 ####
<b>1. <a id="File Digital Signature">计算文件数字签名：File Digital Signature</a></b>   
计算文件/文件夹子文件的MD5、SHA1值，支持计算大文件，支持对文件名的正则过滤，对满足条件的文件进行计算  
![File Digital Signature](./documention/images/File%20Digital%20Signature.png)

<b>2. <a id="Folder And File Operate">文件(夹)查找操作Folder And File Operate</a></b>  
（1）此工具功能丰富，可通过多种条件(名称，类型，时间，大小，其中名称包括后缀名且不区分大小写)查找文件(夹)   
（2）不仅包括文件(夹)查找，也包括复制、剪切、删除文件及删除空文件夹  
（3）通过工具界面可了解其详细功能，注意操作类型非“默认查找”时，需谨慎操作，以防误删除文件  
（4）文件类型配置文件conf\FolderAndFileOperate\filetype.properties  
![Folder And File Operate](./documention/images/Folder%20And%20File%20Operate.png)

<b>3. <a id="JNotify Visual">文件(夹)变化监控：JNotify Visual</a></b>  
使用JNotify监控文件(夹)增删改及重命名  
![JNotify Visual](./documention/images/JNotify%20Visual.png)  
关于JNotify Visual使用的类包<a href="http://improve-lgpl-jars.googlecode.com/files/jnotify-0.94_improve-1.0.jar">jnotify-0.94_improve-1.0.jar</a>，改进自<a href="http://sourceforge.net/projects/jnotify/" target="_blank">jnotify-0.94</a>，项目地址：<a href="http://code.google.com/p/improve-lgpl-jars/" target="_blank">http://code.google.com/p/improve-lgpl-jars/</a>

<b>4. <a id="JODConverter Visual">OpenOffice文档转换：JODConverter Visual</a></b>  
可视化文档转换，支持常见文档的相互转换，如doc/docx转pdf、rtf、text、html，xls/xlsx转pdf、csv、tsv、html，支持的详细转换类型具体参见插件，需Openoffice后台服务支持  
![JODConverter Visual](./documention/images/JODConverter%20Visual.png)
测试OpenOffice版本：3.4  
配置文件：conf/JODConverterVisual/converter.properties  
转换文档：<a href="http://www.artofsolving.com/opensource/jodconverter/guide/supportedformats" target="_blank">Supported Formats</a>&emsp;&nbsp;<a href="http://www.liferay.com/zh/community/wiki/-/wiki/Main/Document+Conversion+with+OpenOffice" target="_blank">Document Conversion with OpenOffice</a>

<b>5. <a id="Text File Split">文本文件切分：Text File Split</a></b>  
按大小或行数切分文件  
![Text File Split](./documention/images/Text%20File%20Split.png)

#### 常用工具集 ####
<b>1. <a id="Code Formatter">Code格式化：Code Formatter</a></b>  
代码格式化，目前仅支持Json  
![Code Formatter](./documention/images/Code%20Formatter.png)  
项目：<a href="https://github.com/google/gson" target="_blank">gson</a>

<b>2. <a id="Password Generator">随机密码生成：Password Generator</a></b>  
使用类库：https://github.com/vt-middleware/passay  
![Password Generator](./documention/images/Password%20Generator.jpg)

<b>3. <a id="QrCode Converter">二维码转换：QrCode Converter</a></b>  
二维码生成、解析  
![QrCode Converter](./documention/images/QrCode%20Converter.png)  
项目：<a href="https://github.com/zxing/zxing" target="_blank">zxing</a>

<b>4. <a id="Regex Tester">正则表达式验证：Regex Tester</a></b>   
常用正则表达式文件conf\RegexTester\expression.properties  
![Regex Tester](./documention/images/Regex%20Tester.png)  
![Regex Tester help](./documention/images/Regex%20Tester-help.png)

<b>5. <a id="Run Script">执行Script脚本：Run Script</a></b>  
此工具简单目前仅可进行简单的计算  
![Run Script](./documention/images/Run%20Script.png)

<b>6. <a id="Time Tool">时间工具：Time Tool</a></b>  
时间字符串与时间戳的相互转换  
![Time Tool](./documention/images/Time%20Tool.png)

<b>7. <a id="ZHConverter Visual">中文简体繁体互转：ZHConverter Visual</a></b>  
![ZHConverter Visual](./documention/images/ZHConverter%20Visual.png)  
项目：<a href="http://code.google.com/p/java-zhconverter/" target="_blank">java-zhconverter</a>

#### 金融工具集 ####
<b>1. <a id="Interest Calculator">利息利率计算器：Interest Calculator</a></b>  
包括：贷款利率计算、等额本金与等额本息比较计算、提前还款计算、分期名义利率真实利率计算  
![Interest Calculator](./documention/images/Interest%20Calculator.jpg)

#### 其他工具集 ####
<b>1. <a id="Color Tool">颜色工具：Color Tool</a></b>  
此工具可进行颜色RGB码与HTML码的相互转换，调色板显示所填写颜色的效果以及选择颜色  
![Color Tool](./documention/images/Color%20Tool.png)

<b>2. <a id="System Information">运行环境信息：System Information</a></b>  
此工具可查看运行机器的Overview(基础信息)、Running Status、System Properties、支持的字符集、字体  
![System Information](./documention/images/System%20Information.png)

#### 过时工具集 ####
<b>1. <a id="Ant Script Auto Build">Ant脚本自动build: Ant Script Auto Build</a></b>  
通过解析Eclipse的Java Project、Dynamic Web Project或者MyEclipse的Web Project的相关配置文件，自动构建者这三类Project的Ant脚本  
![Ant Script Auto Build](./documention/images/Ant%20Script%20Auto%20Build.png)  
Ant Script Auto Build也提供独立版本，下载文件：<a href="https://github.com/baishui2004/common_gui_tools/blob/master/dest/antScriptAutoBuild-1.21.zip?raw=true">antScriptAutoBuild-1.21.zip</a>

<b>2. <a id="Class Finder">Java类查找：Class Finder</a></b>  
查找文件夹下的.class及.java文件，或者文件夹下压缩文件jar,war,aar,ear,zip内的.class及.java文件  
![Class Finder](./documention/images/Class%20Finder.png)  
可配置查找的文件类型（不限于.class及.java文件，可通过配置扩展用于查找其他类型的文件），压缩文件限于java.util.zip.ZipEntry类可解析的类型，配置文件参见conf\ClassFinder\conf.properties

<b>3. <a id="Simple Webview">简易Webview：Simple Webview</a></b>  
说明：Windows系统下可用，Mac及Linux系统下暂不可用  
![Simple Webview](./documention/images/Simple%20Webview.png)  
项目：<a href="https://sourceforge.net/projects/djproject/files/DJ Native Swing/" target="_blank">The DJ Project</a>

release-notes.txt
Version 1.5    2021-3-10
1，完善插件：文件(夹)查找操作(Folder And File Operate)，增加只遍历X级设置；同名文件查询，可选择是否相同后缀名；
           优化重复文件查找速度；
2，完善插件：Code格式化Code Formatter，增加去除json的value，用来对比schema（字符型改空""，数字型改0，换行符统一为\n）；
3，增加插件：利息利率计算器Interest Calculator，包括：贷款利率计算、等额本金与等额本息比较计算、提前还款计算、分期名义利率真实利率计算；
4，增加插件：随机密码生成Password Generator，使用类库：https://github.com/vt-middleware/passay；
5，优化内容：Mac系统上展示程序名称；Mac系统上暂改使用Monaco字体；



Version 1.4    2018-4-22
1，改为Maven项目；
2，删除插件：短网址转换Short URL； 2018-2-14
3，增加插件：二维码转换QrCode Converter； 2018-2-26
4，增加插件：Code格式化Code Formatter； 2018-2-27
5，增加插件：文本文件切分Text File Split； 2018-3-14
6，增加插件：简易Webview Simple Webview； 2018-3-24
7，插件编码转换：Character Converter支持Decode的字符中含有非编码字符； 2018-4-22



Version 1.3    2015-4-22
1，增加插件：时间工具Time Tool； 2015-4-22



Version 1.21    2014-11-27
1，完善插件：文件(夹)查找操作(Folder And File Operate)，增加同名文件查找(比较文件名，不比较后缀名，查找相同文件名称的文件)； 2014-11-27



Version 1.2    2014-8-28
1，增加插件：短网址转换Short URL； 2014-8-28
2，增加插件：颜色工具Color Tool； 2014-8-29
3，增加插件：字符转义工具Escape Character Tool； 2014-8-29



Version 1.1    2013-12-6
1，改善错误日志输出，增加Console日志输出域； 2013-11-22
2，修改插件：OpenOffice文档转换JODConverter Visual，增加“目标文件名是否保留原始文件类型”及“是否作为文本处理”； 2013-11-24
3，增加插件：Java类查找Class Finder，可查找文件夹下的.class及.java文件，以及文件夹下jar、zip中的.class及.java； 2013-12-2
4，修复BUG：在Windows下对无权限访问文件夹进行listFiles()出现NullPointer异常，此时listFiles()==null，解决方法即对此做判断； 2013-12-4
5，修复BUG：Ant脚本自动build插件Ant Script Auto Build，Project的Java Compiler未设置“Enable project specific settings”找不到文件“.settings/org.eclipse.jdt.core.prefs”； 2013-12-6
6，完善插件：Java类查找Class Finder，可配置查找的文件类型（不限于.class及.java文件），压缩文件限于java.util.zip.ZipEntry类可解析的类型； 2013-12-26



Version 1.03   2013-10-1
1，修改JTextArea为自动换行，setLineWrap(true)；  2013-4-28
2，完善插件文件(夹)查找操作(Folder And File Operate)的结果输出统计； 2013-4-30
3，增加插件：中文简体繁体互转ZHConverter Visual； 2013-5-1
4，将需要依赖第三方类包(除commons-codec-*.jar)的插件规划分配到类包bs.util.tool.commongui.plugins.more下。  2013-5-1
5，URL.getPath()处理不了中文或带空格的路径，修改为URLDecoder.decode(path, "UTF-8")（"+"号decode后为空格" "，所以decode前使用"%2b"替换"+"号）； 2013-5-10
6，插件编码转换Character Converter增加手动输入编码转换； 2013-8-8
7，简化精炼属性配置及属性读取； 2013-8-8
9，增加插件的关闭图标，位于插件title栏右侧，左侧是插件图标； 2013-8-8
10，增加插件：文本编码识别JUniversal Chardet； 2013-8-9
11，增加插件：计算文件数字签名File Digital Signature，支持计算大文件； 2013-8-10
12，修改插件：文件(夹)查找操作(Folder And File Operate)，2013-9-25
                    操作类型选择非查找操作的提示框需点击两次才关闭，改为点击一次确定即关闭，代码jComboBox.transferFocus()； 
                    文件显示大小由显示M改增加选择项G、KB、Byte，默认M；
                    重复文件查找方案优化为：取出有相同大小的所有文件，比较文件大小以及前2048Byte的内容的MD5值是否一样，如果两者相同，则认为重复，否则认为不重复，有较小的误差率；
                    增加查找条件：文件/夹路径包含字符，文件/夹路径不包含字符；2013-9-29
13，优化Windows下的启动，使用“start javaw”代替“java”命令以关闭悬停的CMD窗口，增加VM参数“-Xms48m”加快程序的启动；
14，修改插件：编码转换：Character Converter，增加MD5加密的16位值输出(算法：取MD5值32位的第9到24个字符)；2013-9-26
15，修改插件：Ant脚本自动build: Ant Script Auto Build，优化原jar任务为base + jar任务，base及jar任务默认选中(base任务为必选)；2013-9-26
16，优化插件的异常输出以及提炼文件查找的重复代码；2013-9-30
17，将插件文件(夹)变化监控JNotify Visual使用的类包由jnotify-0.93_improve-1.0.jar升级为jnotify-0.94_improve-1.0.jar；2013-9-30
18，增加插件：OpenOffice文档转换JODConverter Visual；2013-10-1



Version 1.02   2013-4-5

1，将配置从类路径下移到与lib同级的conf目录下，将图标从类路径下移到与lib同级的img/icon目录下，方便修改；
2，增加插件：Ant脚本自动build: Ant Script Auto Build，通过解析Eclipse的Java Project、Dynamic Web Project或者MyEclipse的Web Project的相关配置文件，
                                自动构建者这三类Project的Ant脚本。



Version 1.01   2013-3-26

1，解决插件“编码转换：Character Converter”的二进制转换错误BUG，并增加八进制、十进制编码转换。   2013-3-23
2，增加插件：文件(夹)变化监控：JNotify Visual，使用JNotify监控文件(夹)增删改及重命名。
                                                                              注意：文件的剪切(move)操作，Windows下JNotify识别为删除操作，而Deepin 12.06(Linux其他版本待测试)却不识别。



Version 1.0    2013-3-20

Common Gui Tools是用java编写，GUI界面的实用小工具集，暂有6个小工具：

1，编码转换：Character Converter 
2，加密解密：Encrypt And Decrypt 
3，正则表达式验证：Regex Tester
4，执行Script脚本：Run Script
5，文件(夹)查找操作：Folder And File Operate
6，运行环境信息：System Information


src/main/java/bs/tool/commongui/GuiUtils.java
package bs.tool.commongui;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.swing.*;
import java.awt.*;
import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLDecoder;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Map;
import java.util.SortedMap;

public class GuiUtils {

    private static Logger logger = LoggerFactory.getLogger(GuiUtils.class);


    /**************************** 编码 ****************************/

    // 下面两个转码结果一样
    public static String CHARSET_ISO_8859_1 = "ISO-8859-1";
    public static String CHARSET_US_ASCII = "US-ASCII";

    // java转义\ u后面跟的编码, 即Java Unicode转义字符
    public static String CHARSET_UTF_16BE = "UTF-16BE";
    public static String CHARSET_UTF_16LE = "UTF-16LE";

    public static String CHARSET_UTF_8 = "UTF-8";

    // 下面两个转码结果一样
    public static String CHARSET_UTF_16 = "UTF-16";
    public static String CHARSET_Unicode = "Unicode";

    // GB2312 < GBK < GB18030
    public static String CHARSET_GB2312 = "GB2312";
    public static String CHARSET_GBK = "GBK";
    public static String CHARSET_GB18030 = "GB18030";

    public static String CHARSET_Big5 = "Big5";

    /**************************** 算法 ****************************/
    // 可解密/解码算法
    public static String CRYPTO_ASCII = "Ascii";
    public static String CRYPTO_HEX = "Hex";
    public static String CRYPTO_BASE32 = "Base32";
    public static String CRYPTO_BASE64 = "Base64";
    public static String CRYPTO_URL = "URL";
    // 不可解密算法
    public static String CRYPTO_MD5 = "MD5";
    public static String CRYPTO_SHA = "SHA";
    public static String CRYPTO_SHA256 = "SHA256";
    public static String CRYPTO_SHA384 = "SHA384";
    public static String CRYPTO_SHA512 = "SHA512";

    /**************************** 字体 ****************************/

    /**************************** Code类型 ****************************/
    public static String CODE_TYPE_JSON = "JSON";
    public static String CODE_TYPE_PROPERTIES = "Properties";
    public static String CODE_TYPE_YML = "YML";
    public static String CODE_TYPE_XML = "XML";
    public static String CODE_TYPE_JAVA = "Java";
    public static String CODE_TYPE_JS = "JS";
    public static String CODE_TYPE_PYTHON = "Python";
    public static String CODE_TYPE_SQL = "SQL";

    /**************************** Code类型 ****************************/

    /**************************** Split类型 ****************************/
    public static String SPLIT_TYPE_SIZE = "Size(MB)";
    public static String SPLIT_TYPE_LINE = "Line";

    /**************************** Split类型 ****************************/

    /**************************** 密码组合类型 ****************************/
    public static String PASSWORD_COMBINATION_8_CHAR_NUM_SIMPLE = "8位字母数字[简单]";
    public static String PASSWORD_COMBINATION_8_CHAR_NUM_POPULAR = "8位字母数字[复杂]";

    /**************************** 密码组合类型 ****************************/

    /**
     * 所有字体.
     */
    public static Map<String, Font> availableFontsMap = availableFontsMap();

    /**
     * 中文字体集.
     */
    public static String[] fontStyles_cn;
    /**
     * 中文字体.
     */
    public static String fontStyle_cn;
    /**
     * 英文字体集.
     */
    public static String[] fontStyles;
    /**
     * 英文字体.
     */
    public static String fontStyle;
    /**
     * 支持Unicode的字体集.
     */
    public static String[] fontStyles_un;
    /**
     * 支持Unicode的字体.
     */
    public static String fontStyle_un;

    public static Font font12_cn;

    public static Font font13;
    public static Font font13_cn;

    public static Font font14_cn;
    public static Font font14_un;
    public static Font font14b;
    public static Font font14b_cn;

    public static Font font16;

    /**************************** 文件大小单位 ****************************/
    public static String FileSize_PB = "PB";
    public static String FileSize_TB = "TB";
    public static String FileSize_G = "G";
    public static String FileSize_M = "M";
    public static String FileSize_KB = "KB";
    public static String FileSize_Byte = "Byte";

    /**
     * 初始化字体.
     */
    public static void initFont() {
        font12_cn = new Font(fontStyle_cn, Font.PLAIN, 12);

        font13 = new Font(fontStyle, Font.PLAIN, 13);
        font13_cn = new Font(fontStyle_cn, Font.PLAIN, 13);

        font14_cn = new Font(fontStyle_cn, Font.PLAIN, 14);
        font14_un = new Font(fontStyle_un, Font.PLAIN, 14);
        font14b = new Font(fontStyle, Font.BOLD, 14);
        font14b_cn = new Font(fontStyle_cn, Font.BOLD, 14);

        font16 = new Font(fontStyle, Font.PLAIN, 16);
    }

    /**
     * 当前Java虚拟机支持的charset.
     */
    public static SortedMap<String, Charset> availableCharsets() {
        return Charset.availableCharsets();
    }

    /**
     * 支持的字体.
     */
    public static Font[] availableFonts() {
        GraphicsEnvironment environment = GraphicsEnvironment.getLocalGraphicsEnvironment();
        return environment.getAllFonts();
    }

    /**
     * 支持的字体.
     */
    public static Map<String, Font> availableFontsMap() {
        Font[] fonts = availableFonts();
        Map<String, Font> fontsMap = new HashMap<String, Font>();
        for (Font font : fonts) {
            fontsMap.put(font.getFontName(), font);
        }
        return fontsMap;
    }

    /**
     * 按字符集解码.
     */
    public static String encode(String string, String encode, String decode) throws UnsupportedEncodingException {
        return new String(string.getBytes(encode), decode);
    }

    /**
     * 字符串前缀字符填充.
     */
    public static String addFillString(String string, String fill, int interval) {
        StringBuilder sb = new StringBuilder();
        int len = string.length();
        int loop = len / interval;
        for (int i = 0; i < loop; i++) {
            sb.append(fill).append(string.substring(interval * i, interval * (i + 1)));
        }
        if (loop * interval != len) {
            sb.append(fill).append(string.substring(loop * interval, len));
        }
        return sb.toString();
    }

    /**
     * 字符串前填充以满足固定长度.
     */
    public static String fillStringBefore(String string, String fill, int size) {
        StringBuilder sb = new StringBuilder();
        int len = string.length();
        for (int i = 0; i < size - len; i++) {
            sb.append(fill);
        }
        return sb.append(string).toString();
    }

    /**
     * 优先使用前面的字体，如果不存在，则一个一个向后查找..
     */
    public static String getAvailableFont(String[] fontStyles) {
        String fontName = "";
        for (String fontStyle : fontStyles) {
            if (availableFontsMap.containsKey(fontStyle)) {
                fontName = fontStyle;
                break;
            }
        }
        return fontName;
    }

    /**
     * 右填充字符串.
     * <p>
     * <pre>
     * 依据UTF-8编码中文占三个字节, 英文占一个字节, 且宋体下中文占两个半角空格位, 英文占一个半角空格位的特点;
     * 变通填充为字符中有一个中文字符即当两个半角空格位, 一个英文字符即当一个半角空格位.
     * </pre>
     */
    public static String getFillUpString(String string, int size) {
        StringBuilder sb = new StringBuilder(string);
        int len = 0;
        try {
            len = string.length();
            len = len + (string.getBytes("UTF-8").length - len) / 2;
        } catch (UnsupportedEncodingException e) {
            GuiUtils.log(e);
        }
        for (int i = 0; i < size - len; i++) {
            sb.append(" ");
        }
        return sb.toString();
    }

    /**
     * 计算时间, 形式1Days 2Hours 3Minutes 4Seconds.
     */
    public static String getCountTime(long time) {
        time = time / 1000L;
        long minute = 60L;
        long hour = minute * 60L;
        long day = hour * 24L;
        long seconds = time % minute;
        long minutes = (time - seconds) % hour / minute;
        long hours = (time - minutes - seconds) % day / hour;
        long days = (time - hour - minutes - seconds) / day;
        return days + " Days " + hours + " Hours " + minutes + " Minutes " + seconds + " Seconds";
    }

    /**
     * 计算文件(单位)大小，单位为字节Byte.
     */
    public static Double getCountFileSizeUnit(String size, String unit) {
        return getCountFileSizeUnit(size.length() == 0 ? null : Double.parseDouble(size), unit);
    }

    /**
     * 计算文件(单位)大小，单位为字节Byte.
     */
    public static Double getCountFileSizeUnit(Double size, String unit) {
        if (size == null) {
            return null;
        }
        Double bSize = null;
        int cas = 1024;
        if (unit.equals(FileSize_Byte)) {
            bSize = size;
        } else if (unit.equals(FileSize_KB)) {
            bSize = size * cas;
        } else if (unit.equals(FileSize_M)) {
            bSize = size * cas * cas;
        } else if (unit.equals(FileSize_G)) {
            bSize = size * cas * cas * cas;
        } else if (unit.equals(FileSize_TB)) {
            bSize = size * cas * cas * cas * cas;
        } else if (unit.equals(FileSize_PB)) {
            bSize = size * cas * cas * cas * cas * cas;
        }
        return bSize;
    }

    /**
     * 获取目录img/icon目录下的图标.
     */
    public static Icon getIcon(String path, Toolkit kit) {
        return new ImageIcon(getImage(path, kit));
    }

    /**
     * 获取目录img/icon目录下的图片.
     */
    public static Image getImage(String path, Toolkit kit) {
        URL imgURL = null;
        try {
            imgURL = new File(getActualPath(path)).toURI().toURL();
        } catch (MalformedURLException e) {
            GuiUtils.log("Load Image Error.", e);
        }
        return kit.getImage(imgURL);
    }

    /**
     * 获取类加载路径下的图标.
     */
    public static Icon getIconFromClassloader(String classLoaderImagePath, Toolkit kit) {
        return new ImageIcon(getImageFromClassloader(classLoaderImagePath, kit));
    }

    /**
     * 获取类加载路径下的图片.
     */
    public static Image getImageFromClassloader(String classLoaderImagePath, Toolkit kit) {
        // 这种方式可以从jar包中获取资源路径
        URL imgURL = ClassLoader.getSystemResource(classLoaderImagePath);
        return kit.getImage(imgURL);
    }

    /**
     * 获取与Jar lib或classes目录(前提是GuiUtils类被编译在其下)同级的文件(夹)或同级文件夹的子文件(夹)的绝对路径.
     */
    public static String getActualPath(String path) {
        return LIB_PARENT_PATH + path;
    }

    /**
     * Jar lib或classes目录(前提是GuiUtils类被编译在其下)的父路径.
     */
    private static String LIB_PARENT_PATH = getLibParentPath();

    /**
     * 获取Jar lib或classes目录(前提是GuiUtils类被编译在其下)的父路径.
     */
    private static String getLibParentPath() {
        String classResource = GuiUtils.class.getName().replace(".", "/") + ".class";
        String path = "";
        try {
            path = GuiUtils.class.getClassLoader().getResource(classResource).getPath();
            // "+"号decode后为空格" "，"%2b"号decode后为"+"号
            path = path.replace("+", "%2b");
            path = URLDecoder.decode(path, "UTF-8");
            path = path.substring(0, path.length() - classResource.length());
            if (path.contains(".jar!")) {
                path = path.substring(5, path.indexOf(".jar!") + 4);
                path = path.substring(0, path.lastIndexOf("/"));
                path = path.substring(0, path.length() - "lib".length());
            } else {
                path = path.substring(0, path.length() - "classes".length() - 1);
            }
        } catch (UnsupportedEncodingException e) {
            GuiUtils.log(e);
        }
        return path;
    }

    /**
     * 消息输出到控制台.
     *
     * @param msg 消息
     */
    public static void log(String msg) {
        logger.info(msg);
        GuiMain.msgTextArea.append(msg + "\n");
    }

    /**
     * 异常输出到控制台.
     *
     * @param e 异常
     */
    public static void log(Exception e) {
        e.printStackTrace();
        logger.error("", e);
        StackTraceElement[] ste = e.getStackTrace();
        StringBuilder esb = new StringBuilder();
        esb.append("Exception: ").append(e.getClass().getName()).append(": ").append(e.getMessage()).append("\n");
        for (int i = 0; i < 5 && i < ste.length; i++) {
            esb.append("                    ").append(ste[i].toString()).append("\n");
        }
        GuiMain.msgTextArea.append(esb.toString());
    }

    /**
     * 消息、异常输出到控制台.
     *
     * @param msg 消息
     * @param e   异常
     */
    public static void log(String msg, Throwable e) {
        e.printStackTrace();
        logger.error(msg, e);
        log(msg);
        log(e.getMessage());
    }

    /**
     * parseFalse, 有且仅当对象toString值equalsIgnoreCase("false")时返回false, 其他时候返回true.
     */
    public static boolean parseFalse(Object obj) {
        return "false".equalsIgnoreCase(toString(obj)) ? false : true;
    }

    /**
     * trim.
     */
    public static String trim(Object obj) {
        return toString(obj).trim();
    }

    /**
     * toString.
     */
    public static String toString(Object obj) {
        return obj == null ? "" : obj.toString();
    }

    /**
     * is mac.
     */
    public static boolean IS_MAC = System.getProperty("os.name").contains("Mac");

    public static void showAboutMessage(Toolkit kit, JPanel contextPanel, Map<String, String> propsMap) {
        JOptionPane.showMessageDialog(contextPanel, "Version: " + propsMap.get("Version")
                        + "\nAuthor: bs2004@163.com\nDevelop Date: " + propsMap.get("Develop_Date"), "About",
                JOptionPane.INFORMATION_MESSAGE, GuiUtils.getIcon("img/icon/cgt_About.png", kit));
    }

}
